### PR TITLE
Revise cache metrics implementations + tests

### PR DESCRIPTION
**Pull request by OleksiiBondar**
Initially, I was planning just cover package `io.micrometer.core.instrument.binder.cache` with unit tests since there was none. As a result of this effort, I've addressed few bugs/improvements:
* removed non-needed null checks from `CacheMeterBinder` - null values defaulted to 0;
* appended common tags to `cache.remoteSize` metric - can't think of a good reason why it shouldn't be there unless accidentally missed;
* added null check for `objectName` before `MbeanServer.getAttribute` since could be initialized as per implementation (i.e. when `cacheManager` is null)
* used proper methods for
** `cache.put`: `cachePutAddedCount` -> `cachePutUpdatedCount`
** `getLocalOffHeapSize` -> `getLocalOffHeapSizeInBytes`
* add unit test coverage:
Initial coverage
image

After these changes:
image

### CODE REVIEW

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java [hunk 1]
```diff
@@ -67,7 +67,7 @@ public CaffeineCacheMetrics(Cache<?, ?> cache, String cacheName, Iterable<Tag> t
      * @param <C>       The cache type.
      * @return The instrumented cache, unchanged. The original cache is not wrapped or proxied in any way.
      */
-    public static <C extends Cache> C monitor(MeterRegistry registry, C cache, String cacheName, String... tags) {
+    public static <C extends Cache<?, ?>> C monitor(MeterRegistry registry, C cache, String cacheName, String... tags) {
```
Comment (jkschneider) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java]: Could you help me understand in what circumstances this helps? It doesn't help the compiler infer generic type arguments on an expression involving `monitor` is evaluated, and the lack of wildcards doesn't generate compiler warnings about missing type arguments.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java [hunk 2]
```diff
@@ -102,7 +103,7 @@ protected void bindImplementationSpecificMetrics(MeterRegistry registry) {
                 .description("Cache puts resulting in a new key/value pair")
                 .register(registry);
 
-        FunctionCounter.builder("cache.puts.added", stats, StatisticsGateway::cachePutAddedCount)
+        FunctionCounter.builder("cache.puts.added", stats, StatisticsGateway::cachePutUpdatedCount)
```
Comment (jkschneider) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java]: I wonder now if this meter is appropriately named. Perhaps it should be just `cache.puts`, keeping the result tag that indicates the action verb.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/JCacheMetrics.java [hunk 3]
```diff
@@ -129,17 +129,19 @@ protected void bindImplementationSpecificMetrics(MeterRegistry registry) {
...
 
     private Long lookupStatistic(String name) {
-        try {
-            List<MBeanServer> mBeanServers = MBeanServerFactory.findMBeanServer(null);
-            for (MBeanServer mBeanServer : mBeanServers) {
-                try {
-                    return (Long) mBeanServer.getAttribute(objectName, name);
-                } catch (AttributeNotFoundException | InstanceNotFoundException ex) {
-                    // did not find MBean, try the next server
+        if (objectName != null) {
```
Comment (jkschneider) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/JCacheMetrics.java]: This guard looks important.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * @author Oleksii Bondar
+ */
+abstract class AbstractCacheMetricsTest {
+
+    protected abstract Tags getTags();
+
+    /**
+     * Verifies base metrics presense
```
Comment (jkschneider) [inline micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java]: Spelling nit: `presence`

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetricsTest.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Test
+    void doNotReportMetricsForNonLoadingCache() {
+        SimpleMeterRegistry meterRegistry = new SimpleMeterRegistry();
+        Cache<Object, Object> cache = Caffeine.newBuilder().build();
+        CaffeineCacheMetrics metrics = new CaffeineCacheMetrics(cache, "testCache", expectedTag);
+        metrics.bindTo(meterRegistry);
+
+        assertThrows(MeterNotFoundException.class, () -> {
```
Comment (jkschneider) [inline micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetricsTest.java]: You can use `assertThat(meterRegistry.find("cache.load.duration").timeGauge()).isNull())`, which is a bit simpler. `find` is a relaxed version of the stricter `get`. Since the meter won't exist at all, I think it's unnecessary to add the `expectedTag` to the search.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/EhCache2MetricsTest.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    private static CacheManager cacheManager;
+    private static Cache cache;
+
+    private Tags expectedTag = Tags.of("app", "test");
+    private EhCache2Metrics metrics = new EhCache2Metrics(cache, expectedTag);
+
+    @Test
+    void reportMetrics() {
+        SimpleMeterRegistry registry = new SimpleMeterRegistry();
```
Comment (jkschneider) [inline micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/EhCache2MetricsTest.java]: Left-hand side of assignment can be `MeterRegistry`

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.search.RequiredSearch;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+
+/**
+ * @author Oleksii Bondar
+ */
+abstract class AbstractCacheMetricsTest {
+
+    protected abstract Tags getTags();
```
Comment (jkschneider) [inline micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java]: Since the same `app=test` tag is defined on all of the subclasses of `AbstractCacheMetricsTest`, can the setup be simplified?

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/HazelcastCacheMetricsTest.java [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+class HazelcastCacheMetricsTest extends AbstractCacheMetricsTest {
+
+    private static IMap<String, String> cache;
+
+    private Tags expectedTag = Tags.of("app", "test");
+    private HazelcastCacheMetrics metrics = new HazelcastCacheMetrics(cache, expectedTag);
+
+    @Test
+    void reportMetrics() {
+        SimpleMeterRegistry meterRegistry = new SimpleMeterRegistry();
```
Comment (jkschneider) [inline micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/HazelcastCacheMetricsTest.java]: Simplify left-hand assignment to `MeterRegistry`

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java (hunk 1 above)
Comment (OleksiiBondar) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java]: With out parametrization it actually generates compiler warnings
image

Apart from addressing warnings, it also makes implementation consistent across different cache metrics. Here are few examples where it's already parametrized:
https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java#L54
https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastCacheMetrics.java#L62
https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/HazelcastCacheMetrics.java#L47
https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/JCacheMetrics.java#L58
https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/JCacheMetrics.java#L73

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java (hunk 2 above)
Comment (OleksiiBondar) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java]: Good point! I've changed meter name to `cache.puts.updated` and removed qualifying tags since names are unique now.
IMO it's better to have fully qualified names here rather than `cache.puts` since we already have this name defined in https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java#L77
which corresponds to total number of puts per cache not segregated by added/modified flavor.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/JCacheMetrics.java (hunk 3 above)
Comment (OleksiiBondar) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/JCacheMetrics.java]: Yes, even though probability is low there is still a chance of having `cacheManager` as null and it's better to be on the safe side.

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/HazelcastCacheMetricsTest.java (hunk 8 above)
Comment (OleksiiBondar) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/HazelcastCacheMetricsTest.java]: Thanks, addressed across all *CacheMetricsTest classes

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java (hunk 7 above)
Comment (OleksiiBondar) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java]: Right, moved to `AbstractCacheMetricsTest` and removed abstract method :)

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/EhCache2MetricsTest.java (hunk 6 above)
Comment (OleksiiBondar) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/EhCache2MetricsTest.java]: Thanks, addressed across all *CacheMetricsTest classes

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetricsTest.java (hunk 5 above)
Comment (OleksiiBondar) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetricsTest.java]: Thanks, addressed

File: micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java (hunk 4 above)
Comment (OleksiiBondar) [reply micrometer-core/src/test/java/io/micrometer/core/instrument/binder/cache/AbstractCacheMetricsTest.java]: Fixed

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java [hunk 9]
```diff
@@ -47,27 +47,23 @@ public CacheMeterBinder(Object cache, String cacheName, Iterable<Tag> tags) {
 
     @Override
     public final void bindTo(MeterRegistry registry) {
-        if (size() != null) {
```
[the remark this answers is not in the document]
Comment (OleksiiBondar) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java]: Make sense. I've also removed piece of code that defaults to 0 when value is null since we already have null check 
https://github.com/micrometer-metrics/micrometer/blob/master/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java#L52-L55

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java (hunk 2 above)
Comment (jkschneider) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java]: Ah, I see. So it was a name disambiguation concern originally. I don't find `updated` to be any more accurate than `added`, so perhaps we should revert to `added` to preserve compatibility?

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java (hunk 1 above)
Comment (jkschneider) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CaffeineCacheMetrics.java]: Thanks. Ah Eclipse. I think it just isn't smart enough to tell that in this context the generic type parameters don't actually help the compiler. IntelliJ doesn't show the same warnings. I'm happy to leave the type wildcards to help our Eclipse friends.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jkschneider) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java]: I'd like to preserve the null check inside of the gauge as well, for the event that a JCache implementation supports turning stats on and off at runtime. Just extra safety. Want to be really careful that telemetry code never blows up an application.

[the remark this answers is not in the document]
Comment (OleksiiBondar) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java]: I see. For this purpose I'd rather extract value into local variable prior to null evaluation so that we don't need to do double null check. Other aspect of it - it allows to cover code with unit tests, where with double null check it's close to impossible. Let me know if any questions/concerns.

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java (hunk 2 above)
Comment (OleksiiBondar) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/EhCache2Metrics.java]: Done

File: micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java [hunk 10]
```diff
@@ -47,23 +47,17 @@ public CacheMeterBinder(Object cache, String cacheName, Iterable<Tag> tags) {
...
-            Gauge.builder("cache.size", cache.get(),
-                    c -> {
-                        Long size = size();
-                        return size == null ? 0 : size;
-                    })
+        Long cacheSize = size();
```
Comment (jkschneider) [inline micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java]: This introduces a bug, causing the initial cache size to be reported from this point forward.

Comment (OleksiiBondar) [reply micrometer-core/src/main/java/io/micrometer/core/instrument/binder/cache/CacheMeterBinder.java]: (facepalm) Thank you for catching this! Addressed
